### DISCUSSION TITLE
View or decode raw buffer

### DISCUSSION THREAD

**Discussion by tewarid** [raspberrypi/picamera2#488] [Q&A]
I am using code such as the following to save full resolution raw frame buffer. I am expecting it to be in SGBRG10_CSI2P format but I just realized I don't know how to decode it (let's say using opencv2) or a program that may be able to view it. Any help is much appreciated.

```python
from picamera2 import Picamera2, Preview
import time

picam2 = Picamera2()
camera_config = picam2.create_still_configuration(raw={})
picam2.configure(camera_config)
picam2.start()
time.sleep(2)
counter = 0
start = time.time()
while 1:
    print(f"capture {counter}")
    (buffer, ), metadata = picam2.capture_buffers(["raw"])
    buffer.tofile(f"test{counter}.raw")
    counter += 1
    if counter == 10:
        break
end = time.time()
print(end - start)
```

**Comment by davidplowman** [comment]
Hi, it of course depends exactly what result you want. If you want a 2-dimensional array of 16-bit values where each value is a single value (with 10 bits of range) from the raw Bayer array:

* In the raw stream ask for a format of `SBGGR10` (instead of `SBGGR10_CSI2P`), this will give you unpacked values and save you having to do all the bit operations yourself
* Use `capture_array` instead of `capture_buffer` to get a 2-d array
* The use `numpy.view` to view this as an array of 16-bit values.

An alternative might be simply to save a DNG. Then you have lots of 3rd party software that will read those, but it may not be what you want.
